src/api/customization.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This file contains all customizable constants.
+//!
+//! If you adapt them, make sure to run the tests before flashing the firmware.
+//! Our deploy script enforces the invariants.
+
+pub trait Customization {
+    /// Maximum message size send for CTAP commands.
+    ///
+    /// The maximum value is 7609, as HID packets can not encode longer messages.
+    /// 1024 is the default mentioned in the authenticatorLargeBlobs commands.
+    /// Larger values are preferred, as that allows more parameters in commands.
+    /// If long commands are too unreliable on your hardware, consider decreasing
+    /// this value.
+    fn max_msg_size(&self) -> usize;
+}
+
+#[derive(Clone)]
+pub struct CustomizationImpl {
+    pub max_msg_size: usize,
+}
+
+pub const DEFAULT_CUSTOMIZATION: CustomizationImpl = CustomizationImpl { max_msg_size: 7609 };
+
+impl Customization for CustomizationImpl {
+    fn max_msg_size(&self) -> usize {
+        self.max_msg_size
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    #[allow(clippy::assertions_on_constants)]
+    fn test_invariants() {
+        assert!(DEFAULT_CUSTOMIZATION.max_msg_size >= 1024);
+        assert!(DEFAULT_CUSTOMIZATION.max_msg_size <= 7609);
```

Comment:
Let's instead have a validation function if those invariants should hold for all customizations.

```rust
fn is_valid(customization: &impl Customization) -> bool {
    // explanation
    if ... { return false; }
    // explanation
    if ... { return false; }
    true
}
```

Then the test would just `assert!(is_valid(&DEFAULT_CUSTOMIZATION))`.

Reply:
We can probably compile that function for `std` only?

Reply:
What would be the motivation? I can see 3:
- Reduce compilation time. I'm not sure of the impact and if it's needed.
- Prevent users to use it outside tests. Makes sense. But not sure why this should be forbidden.
- Reduce binary size. This shouldn't be a concern if the function is not used.

Reply:
It conveys the intention that all of these are constants, so they should be tested during compile time or tests, not runtime.

What we initially wanted was static assertions, but IIRC that wasn't well supported back then. Now we have an additional use case, but runtime testing in prod is still a bad idea.

Reply:
Actually, we should compile that for cfg(test) only otherwise there will be dead code warning

Reply:
That's weird if we get a dead-code warning. If it's a public function it should be fine. We might use it in fuzzing.

Sounds good for the `std` gate.

Reply:
Done